File: Backend/models/llm.py
from sqlalchemy import Column, ForeignKey, Integer, String, Text, DateTime
from sqlalchemy.orm import relationship
from datetime import datetime
from config.database import Base


class LLM(Base):
    __tablename__ = "llm"

    id = Column(Integer, primary_key=True, autoincrement=True)
    prompt = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.now)
    company_id = Column(Integer, ForeignKey("company.id"), nullable=True)
    system_greeting = Column(Text, nullable=True)
    botName = Column(String(100), nullable=True)
    
    
    chunksize = Column(Integer, nullable=True)
    chunkoverlap = Column(Integer, nullable=True)
    topk = Column(Integer, nullable=True)
    
    
    bot_model_detail_id = Column(Integer, nullable=True)  # ID của llm_detail dùng cho bot (1=gemini, 2=gpt)
    embedding_model_detail_id = Column(Integer, nullable=True)  # ID của llm_detail dùng cho embedding

    # Quan hệ 1-nhiều với LLMDetail
    llm_details = relationship("LLMDetail", back_populates="llm", cascade="all, delete-orphan")


class LLMDetail(Base):
    __tablename__ = "llm_detail"

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(150), nullable=False)  # "gemini" hoặc "gpt"
    key_free = Column(Text, nullable=True)
    llm_id = Column(Integer, ForeignKey("llm.id"), nullable=False)

    llm = relationship("LLM", back_populates="llm_details")

    
    llm_keys = relationship("LLMKey", back_populates="llm_detail", cascade="all, delete-orphan")


class LLMKey(Base):
    __tablename__ = "llm_key"

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String(150), nullable=False)
    key = Column(String(150), nullable=False)
    type = Column(String(50), nullable=False, default="bot")  # "bot" hoặc "embedding"
    created_at = Column(DateTime, default=datetime.now)
    updated_at = Column(DateTime, default=datetime.now, onupdate=datetime.now)
    llm_detail_id = Column(Integer, ForeignKey("llm_detail.id"), nullable=False)

    # Quan hệ ngược lại với LLMDetail
    llm_detail = relationship("LLMDetail", back_populates="llm_keys")

File: Frontend/src/types/llm.ts
// Interface cho LLM Key data (không phải response từ API)
export interface LLMKey {
  name: string;
  key: string;
  type: "bot" | "embedding";
}

// Interface cho LLM Key từ API response
export interface LLMKeyData {
  id: number;
  name: string;
  key: string;
  type: "bot" | "embedding";
  llm_detail_id: number;
  created_at: string;
  updated_at: string;
}

// Interface cho LLM Detail từ API response
export interface LLMDetailResponse {
  id: number;
  name: string;
  key_free: string;
  llm_keys: LLMKeyData[];
}

// Interface cho LLM data (không phải response từ API)
export interface LLM {
  prompt: string;
  system_greeting: string;
  botName: string;
  bot_model_detail_id: string;
  embedding_model_detail_id: string;
  company_id: string;
  chunksize?: number;
  chunkoverlap?: number;
  topk?: number;
}

// Interface cho LLM data từ API response
export interface LLMData {
  id: number;
  prompt: string;
  created_at: string;
  system_greeting: string;
  botName: string;
  bot_model_detail_id: number;
  embedding_model_detail_id: number;
  chunksize?: number;
  chunkoverlap?: number;
  topk?: number;
  llm_details: LLMDetailResponse[];
}

// Interface cho LLM response từ API
export interface LLMResponse {
  message: string;
  llm: LLMData;
}

// Interface cho LLM Key response từ API
export interface LLMKeyResponse {
  message: string;
  llm_key: LLMKeyData;
}
